cpp/src/arrow/builder.h

```diff
@@ -315,7 +327,11 @@ inline uint8_t expanded_uint_size(uint64_t val, uint8_t current_int_size) {
 
 class ARROW_EXPORT AdaptiveUIntBuilder : public AdaptiveIntBuilderBase {
  public:
+#ifdef ARROW_NO_DEFAULT_MEMORY_POOL
   explicit AdaptiveUIntBuilder(MemoryPool* pool);
```

Comment:
I would rather prefer a second internal macro:

```
#ifdef ARROW_NO_DEFAULT_MEMORY_POOL
#define ARROW_ARG_MEMORY_POOL MemoryPool* pool
#else
#define ARROW_ARG_MEMORY_POOL MemoryPool* pool = default_memory_pool()
#endif
```
This would avoid the `if…else…endif` duplications here.